tests/ga/test_remoteaccess_handler.py

```diff
@@ -0,0 +1,358 @@
+
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="]aPPEv}uNg1FPnl?")
+    def test_add_user_already_existing(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = "]aPPEv}uNg1FPnl?"
+        tstuser = "foobar"
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        pwd = tstpassword
+        rah.add_user(tstuser, pwd, expiration_date)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertTrue(tstuser in os_util.all_users, "{0} missing from users".format(tstuser))
+        self.assertEqual(1, os_util.all_users.keys().__len__())
+        actual_user = os_util.all_users[tstuser]
+        self.assertEqual(actual_user[7], (expiration_date + timedelta(days=1)).strftime("%Y-%m-%d"))
+        # add the new duplicate user, ensure it's not created and does not overwrite the existing user.
+        # this does not test the user add function as that's mocked, it tests processing skips the remaining
+        # calls after the initial failure
+        new_user_expiration = datetime.utcnow() + timedelta(days=5)
+        rah.add_user(tstuser, pwd, new_user_expiration)
+        self.assertTrue(tstuser in os_util.all_users, "{0} missing from users after dup user attempted".format(tstuser))
+        self.assertEqual(1, os_util.all_users.keys().__len__())
+        actual_user = os_util.all_users[tstuser]
+        self.assertEqual(actual_user[7], (expiration_date + timedelta(days=1)).strftime("%Y-%m-%d"))
+
+    # delete_user tests
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="]aPPEv}uNg1FPnl?")    
+    def test_delete_user(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = "]aPPEv}uNg1FPnl?"
+        tstuser = "foobar"
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        pwd = tstpassword
+        rah.add_user(tstuser, pwd, expiration_date)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertTrue(tstuser in os_util.all_users, "{0} missing from users".format(tstuser))
+        rah.delete_user(tstuser)
+        self.assertFalse(tstuser in os_util.all_users)
+
+    def test_handle_failed_create_with_bad_data(self):
+        mockOSUtil = MockOSUtil()
+        testusr = "foobar"
+        mockOSUtil.all_users[testusr] = (testusr, None, None, None, None, None, None, None)
+        rah = RemoteAccessHandler()
+        rah.os_util = mockOSUtil
+        rah.handle_failed_create("", "test message")
+        self.assertEqual(1, mockOSUtil.all_users.keys().__len__())
+        self.assertTrue(testusr in mockOSUtil.all_users, "Expected user {0} missing".format(testusr))
```

Comment:
Because you're not mocking the logger, you can't check to see if appropriate error messages are being logged. You should add that mocking and do those checks.

Reply:
added

---
In reply to: [194538421](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r194538421) [](ancestors = 194538421)